stdlib/domain.ml

```diff
@@ -87,17 +87,33 @@ let do_before_first_spawn () =
     first_spawn_function := (fun () -> ())
   end
 
-let at_exit_key = CamlinternalTLS.new_key (fun () -> (fun () -> ()))
+module IMap = Map.Make (struct type t = int let compare = (-) end)
+let at_exit_callbacks = ref IMap.empty
+let at_exit_mutex = Mutex.create () (* protects [at_exit_callbacks] *)
 
 let at_exit f =
-  let old_exit : unit -> unit = CamlinternalTLS.get at_exit_key in
-  let new_exit () =
-    f (); old_exit ()
-  in
-  CamlinternalTLS.set at_exit_key new_exit
+  let domain_id = (self () :> int) in
+  Mutex.protect at_exit_mutex (fun () ->
+    let m = !at_exit_callbacks in
+    let old_exit_opt = IMap.find_opt domain_id m in
+    let new_exit = match old_exit_opt with
+    | None -> f
+    | Some old_exit -> (fun () -> (f (); old_exit ()))
+    in
+    at_exit_callbacks := IMap.add domain_id new_exit m)
```

Comment:
We do not know much about the future usages of `Domain.at_exit` so we don't know whether the mutex is expensive. Do note, though, that it is fairly straightforward to avoid holding the mutex in most cases (more details on request).

Minor detail, this change fixes a thread-unsafety of Domain.at_exit of current trunk.

Reply:
Thanks. Does the thread-unsafety of `Domain.at_exit` have an issue on ocaml/ocaml?

> Do note, though, that it is fairly straightforward to avoid holding the mutex in most cases (more details on request).

If it is very straightforward, I am happy to hear. With that said, do you think we can have it as a separate PR? We've spent quite a bit of time on this PR and #12719 already. I'm keen to get this done and dusted as a first cut.

Reply:
> Thanks. Does the thread-unsafety of `Domain.at_exit` have an issue on ocaml/ocaml?

Not that I know of, I just noticed it while reviewing.

> If it is very straightforward, I am happy to hear. With that said, do you think we can have it as a separate PR? We've spent quite a bit of time on this PR and #12719 already. I'm keen to get this done and dusted as a first cut.

Sure, at the same time we can check if we need a "yet to run" flag as for Stdlib.at_exit. In this case I think this commit is good to go.

Reply:
> we can check if we need a "yet to run" flag as for Stdlib.at_exit. In this case I think this commit is good to go.
sounds good.